File: backend/app/schemas/signal.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import ScrapeSourceType, SignalPriority, SignalStatus, SignalType


class SignalResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    scrape_job_id: UUID
    scrape_result_id: UUID
    signal_type: SignalType
    source_type: ScrapeSourceType
    title: str
    summary: str | None
    raw_snippet: str | None
    source_url: str | None
    source_label: str | None
    extracted_data: dict[str, Any]
    confidence_score: float | None
    status: SignalStatus
    priority: SignalPriority | None
    agent_metadata: dict[str, Any] = Field(default_factory=dict)
    reviewed_by_id: UUID | None
    reviewed_at: datetime | None
    dismissed_reason: str | None
    converted_at: datetime | None
    extraction_model: str | None
    detected_at: datetime | None
    created_at: datetime
    updated_at: datetime


class SignalListResponse(BaseModel):
    items: list[SignalResponse] = Field(default_factory=list)
    total: int
    limit: int
    offset: int

File: frontend/src/types/signal.ts
import type { SignalPriority } from '@/types/lead'

export type SignalStatus =
  | 'new'
  | 'reviewed'
  | 'qualified'
  | 'dismissed'
  | 'converted'
  | 'duplicate'

export type SignalType =
  | 'lead'
  | 'competitor_intel'
  | 'review'
  | 'mention'
  | 'hiring'
  | 'pricing_change'
  | 'news'
  | 'other'

export type ScrapeSourceType = 'web' | 'social' | 'competitor' | 'review' | 'other'

export interface Signal {
  id: string
  organization_id: string
  scrape_job_id: string
  scrape_result_id: string
  signal_type: SignalType
  source_type: ScrapeSourceType
  title: string
  summary: string | null
  raw_snippet: string | null
  source_url: string | null
  source_label: string | null
  extracted_data: Record<string, unknown>
  confidence_score: number | null
  status: SignalStatus
  priority: SignalPriority | null
  agent_metadata: Record<string, unknown>
  reviewed_by_id: string | null
  reviewed_at: string | null
  dismissed_reason: string | null
  converted_at: string | null
  extraction_model: string | null
  detected_at: string | null
  created_at: string
  updated_at: string
}

export interface PaginatedSignalsResponse {
  items: Signal[]
  total: number
  limit: number
  offset: number
}

export interface SignalListParams {
  status?: SignalStatus
  signal_type?: SignalType
  source_type?: ScrapeSourceType
  scrape_job_id?: string
  scrape_result_id?: string
  limit?: number
  offset?: number
}

export interface SignalUpdateRequest {
  status?: SignalStatus
  priority?: SignalPriority | null
  dismissed_reason?: string | null
}

export type SignalSortField =
  | 'title'
  | 'lead_score'
  | 'status'
  | 'signal_type'
  | 'source_label'
  | 'priority'
  | 'detected_at'

export type SortDirection = 'asc' | 'desc'
